File: examples/性能测试/bench_bitwise_loop.c
/* bench_bitwise_loop.c — bench_bitwise_loop.leno 的 C 对照实现
 *
 * 语义对齐：Leno 的 int 是 int48（有符号 48 位），这里用 int64_t。本基准里值始终
 * 落在 0..1200（>>8 后 &7 夹到 0..7，再 ^1125 &1135 +77），所以位宽不影响结果，
 * 也不会有溢出/提升 BigInt 的差异。
 *
 * 编译/运行（mingw-w64）：
 *   gcc -O2 -o bench_bitwise_loop_c.exe bench_bitwise_loop.c ; bench_bitwise_loop_c.exe
 * 计时用 QueryPerformanceCounter（与 Leno 的 times.ms() 一样是墙钟毫秒）。
 */
#include <stdio.h>
#include <windows.h>

int main(void) {
    long long inp = 12345;
    const long long iterations = 2000000000LL;
    LARGE_INTEGER freq, t0, t1;
    long long i;

    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&t0);

    for (i = 0; i < iterations; i++) {
        inp = inp >> 8;
        inp = inp & 7;
        inp = 1125 ^ inp;
        inp = inp & 1135;
        inp = inp + 77;
    }

    QueryPerformanceCounter(&t1);
    double ms = (double)(t1.QuadPart - t0.QuadPart) * 1000.0 / (double)freq.QuadPart;

    printf("结果: %lld\n", inp);
    printf("耗时: %.0f ms\n", ms);
    return 0;
}

File: examples/性能测试/bench_bitwise_loop.lua
-- bench_bitwise_loop.lua —— 与 bench_bitwise_loop.leno 同口径的位运算循环
--   （对齐 Java 的位运算跑分测试：移位 / 与 / 异或 / 与 / 加，每轮 5 个整数操作）
--
-- 迭代次数取 **2 亿**：20 亿在 Lua 侧要跑一百多秒，对表没必要。
-- 轮数口径与 .leno 的 `for 0 : iterations to i` 严格对齐 —— Leno 的 `:` 是**闭区间**，
--   所以那边是 iterations + 1 轮；Lua 的数值 for `for i = 0, N` 同样是闭区间 ⇒ 两边一致。
--
-- ⚠ Lua 的位运算符与 C 不同：**异或是 `~`（不是 `^`）**，`^` 在 Lua 里是乘方。
--   `>>` 在 Lua 是逻辑右移、Leno 是算术右移 —— 本基准的 inp 全程为正，两者一致。

local inp = 12345
local iterations = 200000000  -- 2 亿次（含下标 0 ⇒ 实际 200000001 轮）

local start = os.clock() * 1000  -- 与其它 Lua 基准同口径：CPU 时间 × 1000 = 毫秒

for i = 0, iterations do
    inp = inp >> 8
    inp = inp & 7
    inp = 1125 ~ inp
    inp = inp & 1135
    inp = inp + 77
end

local finish = os.clock() * 1000
print("结果: " .. inp)
print("耗时: " .. (finish - start) .. " ms")
